Add monitoring for new channels

edmangimelli opened the pull request.

edmangimelli committed:
Add monitoring for new channels

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+	if err != nil {
+		return err
+	}
+	current := make([]ChannelAndID, 0, 50)
```
Setting length to 0 and capacity to 50 allows us to avoid the performance hit from the first 50 calls to `append`. This can be useful in some cases but it's unnecessary here. Set the length to `len(rawCurrent)` and don't use `append`. You can do a `for i, v := range rawCurrent` and use the `i` to write to `current[i]` directly. No copying, no allocation, and no overhead of calling `append`.

edmangimelli replied:
Whoops! I feel dumb for missing this one.

jcbwlkr replied:
No need to feel dumb!

edmangimelli replied:
:D

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+	params := slack.NewPostMessageParameters()
+	params.AsUser = true
+	params.EscapeText = false
+	msg := fmt.Sprintf("Channel changes:\n%s", diff)
```
I'm good with just using the output of `cmp.Diff` because it's *good enough*. A real easy way we could make it more legible is to surround the `%s` in ``` blocks

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+func writeChannelsToColl(coll *mgo.Collection, chans []ChannelAndID) error {
+	coll.DropCollection()
+	for _, v := range chans {
+		if err := coll.Insert(v); err != nil {
```
Instead of inserting each value one at a time in a loop you can do `coll.Insert(chans...)` and they will be unpacked to pass in to the variadic function Insert

edmangimelli replied:
Cool --didn't research that method well enough.

jcbwlkr replied:
So as we discovered that doesn't actually work because Go won't automatically convert `[]Channel` to `[]interface{}`. What you have here is fine.

edmangimelli replied:
:+1:

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+
+
+func writeChannelsToColl(coll *mgo.Collection, chans []ChannelAndID) error {
+	coll.DropCollection()
```
Can this error?

edmangimelli replied:
whoops. yes.

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+	"gopkg.in/mgo.v2"
+)
+
+type ChannelAndID struct {
```
We don't need to make a new type for this. Use the existing `type Channel` and add an `ID` field to it.

edmangimelli replied:
Yeah, it didn't feel right making that new type, but, for some reason, it felt wrong to modify the existing type `Channel`. I'll fix it.

jcbwlkr commented on channelDiff.go:
```diff
@@ -0,0 +1,89 @@
+
+import (
+	"log"
+	"fmt"
```
Put a blank line after this import. I prefer to organize imports like

```
import (
    "std/lib/stuff"
    "std/lib/stuff"

    "3rd/party/stuff"
    "3rd/party/stuff"

    "our/internal/packages"
    "our/internal/packages"
)
```

edmangimelli replied:
I like it!

jcbwlkr reviewed (commented):
Sweet! Thanks for doing this. It looks great. I suggest a few minor change and one that might be a bit more work: use the existing `type Channel` instead of `type ChannelAndID`

edmangimelli committed:
Fix monitoring for new channels

Fixed 5 minor issues with pull request. These issues were
highlighted in the comments to the pull request. The only
issue that wasn't resolved was:

func writeChannelsToColl(coll *mgo.Collection, chans []ChannelAndID) error {
	coll.DropCollection()
	for _, v := range chans {
		if err := coll.Insert(v); err != nil {

The for loop was left in because Go will not convert a
[]T []Interface{} for you
https://stackoverflow.com/a/12754757/859353
(as pointed out by J.W.)

jcbwlkr committed:
Remove some whitespace

jcbwlkr committed:
Simplify some error handling
